### ISSUE TITLE
Should default cache_dir to $XDG_CACHE_HOME/ccache

### ISSUE THREAD

**Issue by antoniojpfernandes** [ccache/ccache#191]
Problems using cache_dir=$HOME/.ccache: it clutters the home dir when showing hidden files. Also, it slows down utilities scanning the home dir.

Advantages using cache_dir=$XDG_CACHE_HOME/ccache: one less dot-dir in the home directory. Also, makes it easy to delete all cache at once. $XDG_CACHE_HOME/ccache will tipically be $HOME/.cache/ccache

Of course, the configuration file is not cache, so keep reading .ccache/ccache.conf if existing, but favor this as new location: $XDG_CONFIG_HOME/ccache.conf 

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

(I know I can configure ccache to use whatever cache_dir I want. The proposal is about setting a better default for everyone.)

**Comment by jrosdahl** [comment]
Thanks for your interest in ccache!

Using XDG_CACHE_HOME was discussed and rejected a couple of years ago, see https://bugzilla.samba.org/show_bug.cgi?id=10383. Did you see that discussion before writing this issue?

**Comment by antoniojpfernandes** [comment]
Thank you for the polite reply. I looked for duplicate discussions here on github but did not know about the bugzilla instance. I've read the previous discussion now and I see you have thought deeply about it already.

Indeed, storing configuration inside $XDG_CACHE_HOME is wrong and not an option. The standard offers $XDG_CONFIG_DIR fo that; quoting you:

> The configuration would have to be stored in XDG_CONFIG_DIR, but reading configuration from XDG_CONFIG_DIR when CCACHE_DIR is set to a shared directory does not make sense.

You are right. My suggetion is this: if CCACHE_DIR is set (either as an envar or in the system configuration file), it should override BOTH the default cache dir and the default config path (but the later can be further overiden by CCACHE_CONFIGPATH, of course!).

The documentation would then explain that, if neither the CCACHE_DIR envar nor the cache_dir configuration option are set, ccache will take cache and user configuration from separate standard directories, BUT, if either CCACHE_DIR or cache_dir are set, ccache will take BOTH cache and configuration from the same custom directory.

Do you think this makes sense? It offers: 1) full compatibility for current schema (user can set cache_dir=$HOME/.ccache/ and everything will remain the same), 2) same configuration power as today, 3) a nicer default, and 4) compliance with standards.

**Comment by jrosdahl** [comment]
> Do you think this makes sense?

Sounds like a reasonable idea. Thanks!

Here are some downsides I can think of:

1. It introduces more complexity and more to explain in the manual.
2. It will silently change the default so that users' old `~/.ccache` directory no longer will be used – their old settings won't be saved and no existing cache results will be used.
3. It won't be possible to configure compatibility for the current scheme since variable references (e.g. `$HOME`) are not expanded in values of configuration settings. But it would of course be possible to add support for that.

A modification of the idea that solves downsides 2 and 3 is to introduce a special value of cache_dir that opts in on the XDG layout. So if e.g. `CCACHE_DIR=xdg` or `cache_dir = xdg` in the configuration, then do it the XDG way. What do you think?

**Comment by Earnestly** [comment]
The traditional way in which migrations are handled is to always look for the original path and use that before attempting to follow the XDG convention.

This article was written to provide a bunch of case studies so that developers might see how other projects have handled it: https://wiki.archlinux.org/index.php/XDG_Base_Directory_support

It's not perfect, but better than the free-for-all we had before.

**Comment by jrosdahl** [comment]
> The traditional way in which migrations are handled is to always look for the original path (or environments) and use that before attempting to follow the XDG convention.

Sounds reasonable.


So to sum it up, here's how we could select the cache-specific configuration file (`primary_config_path` in the pseudo code below) and cache directory to use:

```python
if CCACHE_CONFPATH is set:
    primary_config_path = "$CCACHE_CONFPATH"
elif CCACHE_DIR is set:
    primary_config_path = "$CCACHE_DIR/ccache.conf"
elif "$HOME/.ccache" exists:
    primary_config_path = "$HOME/.ccache/ccache.conf"
elif XDG_CONFIG_HOME is set:
    primary_config_path = "$XDG_CONFIG_HOME/ccache/ccache.conf"
else:
    primary_config_path = "$HOME/.config/ccache/ccache.conf"

conf = conf_create()
if not CCACHE_CONFPATH is set:
    conf_read(conf, secondary_config_path)
conf_read(conf, primary_config_path)
conf_update_from_environment(config)

if config.cache_dir is not set:
    if "$HOME/.ccache" exists:
        config.cache_dir = "$HOME/.ccache"
    elif XDG_CACHE_HOME is set:
        config.cache_dir = "$XDG_CACHE_HOME/ccache"
    else:
        config.cache_dir = "$HOME/.cache/ccache"
```

(See [`initialize()` in `ccache.c`](https://github.com/ccache/ccache/blob/284e3a0c66193639171ff2a6ec4ff00a4a90dea9/src/ccache.c#L3150) for the actual current logic.)

I think that a tricky part of this will be to document the semantics in a way that is easy to understand...

**Comment by antoniojpfernandes** [comment]
> I think that a tricky part of this will be to document the semantics in a way that is easy to understand...

That's true, ccache documentation emphasizes 2 levels of configuration: "system-wide" and "cache-specific". And "cache-specific " doesn't map well to $XDG_CONFIG_HOME which is the place for "user-wide" configuration. The same user may have multiple caches, and it would be weird for $XDG_CONFIG_HOME/ccache/ccache.config to apply to only one of them.

So, maybe this calls for a new level of configuration, with priority between system-wide and cache-specific? Like this (modifications in bold):

> ## Configuration
> ccache’s default behavior can be overridden by configuration file settings, which in turn can be overridden by environment variables with names starting with CCACHE_. ccache normally reads configuration from two **or three** files: first a system-level configuration file, secondly **a user-level configuration, and thirdly** a cache-specific configuration file. The priority of configuration settings is as follows (where 1 is highest): 
> 1. Environment variables.
> 2. The cache-specific configuration file, if applicable. This is **$CCACHE_DIR/ccache.conf, if ccache_dir is set, otherwise $HOME/.ccache/ccache.conf if existing. The generic cache $XDG_CACHE_HOME/ccache does not support cache-specific settings.**
> 3. **The per-user configuration file $XDG_CACHE_HOME/ccache.conf (typically $XDG_CONFIG_HOME/ccache/ccache.conf).**
> 4. The system-wide configuration file <sysconfdir>/ccache.conf (typically /etc/ccache.conf or /usr/local/etc/ccache.conf).
> 5. Compile-time defaults.

[...]

>  ### cache_dir (CCACHE_DIR)
>
>  This setting specifies a custom directory for ccache to keep its cached compiler outputs **and cache-specific configuration file.** It will only take effect if set in the system-wide **or user-wide** configuration file, or as an environment variable. **If not set, ccache will use *$HOME/.ccache*, if existing, for backwards compatibility. Otherwise, ccache uses the generic cache at $XDG_CACHE_HOME/.ccache and takes no cache-specific configuration file.**

Of course, what this would mean in pseudocode is:

```python
if CCACHE_CONFPATH is set:
    primary_config_path = "$CCACHE_CONFPATH"
elif CCACHE_DIR is set:
    primary_config_path = "$CCACHE_DIR/ccache.conf"
elif "$HOME/.ccache" exists:
    primary_config_path = "$HOME/.ccache/ccache.conf"
else
    primary_config_path = ""

if XDG_CONFIG_HOME is set:
    secondary_config_path = "$XDG_CONFIG_HOME/ccache/ccache.conf"
else:
    secondary_config_path = "$HOME/.config/ccache/ccache.conf"

conf = conf_create()
if not CCACHE_CONFPATH is set:
    conf_read(conf, secondary_config_path)
    conf_read(conf, tertiary_config_path)
conf_read(conf, primary_config_path)
conf_update_from_environment(config)

if config.cache_dir is not set:
    if "$HOME/.ccache" exists:
        config.cache_dir = "$HOME/.ccache"
    elif XDG_CACHE_HOME is set:
        config.cache_dir = "$XDG_CACHE_HOME/ccache"
    else:
        config.cache_dir = "$HOME/.cache/ccache"
```

**Comment by antoniojpfernandes** [comment]
A simpler variation of the previous suggestion is not to use "$XDG_CONFIG_HOME/ccache/ccache.conf" at all, but simply state that cache-specific configuration is only applicable if ccache_dir is set or (for backwards compatibility) if "$HOME/.ccache" exists.

**Comment by afbjorklund** [comment]
Naming above is a little confusing, since ccache printout uses "secondary config" to refer to the system-wide config. I suppose that is now known as "tertiary config", under the suggested extended scheme ?

Would probably be less confusing to either keep the old, or go "all in" on the XDG directories:

Old setup:
```
cache directory                     /home/anders/.ccache
primary config                      /home/anders/.ccache/ccache.conf
secondary config      (readonly)    /usr/local/etc/ccache.conf
```
New setup:
```
cache directory                     /home/anders/.cache/ccache
primary config                      /home/anders/.config/ccache/ccache.conf
secondary config      (readonly)    /usr/local/etc/ccache.conf
```

Which reminds me, I suppose that CCACHE_TEMPDIR should *not* be `~/.cache/ccache/tmp` ?

**Comment by jrosdahl** [comment]
Thanks for your input.

@antoniojpfernandes wrote:

> That's true, ccache documentation emphasizes 2 levels of configuration: "system-wide" and "cache-specific". And "cache-specific " doesn't map well to $XDG_CONFIG_HOME which is the place for "user-wide" configuration. The same user may have multiple caches, and it would be weird for $XDG_CONFIG_HOME/ccache/ccache.config to apply to only one of them.
>
> So, maybe this calls for a new level of configuration, with priority between system-wide and cache-specific? [...] The generic cache $XDG_CACHE_HOME/ccache does not support cache-specific settings.

I don't see how introducing a third config in that way would work without breaking backward compatibility. The thing is that the primary config needs to be writable since `ccache -M`, `ccache -o`, etc write to it, so it has to be associated with just one cache.

I intended my suggested pseudo code to tie `$XDG_CONFIG_HOME/ccache` and `$XDG_CACHE_HOME/ccache` together so that if you set an explicit cache dir (either via environment or via the secondary config), `$XDG_CONFIG_HOME/ccache/ccache.conf` wouldn't be used at all. But I see now that my pseudo code isn't quite right. New attempt:

```python
conf = conf_create()

if CCACHE_CONFIGPATH is set:
    # Note: CCACHE_CONFIGPATH is mainly (only?) used by ccache's own test suite
    primary_config_path = "$CCACHE_CONFIGPATH"
else:
    conf_read(conf, secondary_config_path)
    if CCACHE_DIR is set:
        primary_config_dir = "$CCACHE_DIR"
    elif conf.cache_dir is set:  # from secondary config
        primary_config_dir = conf.cache_dir
    elif "$HOME/.ccache" exists:
        primary_config_dir = "$HOME/.ccache"
    elif XDG_CONFIG_HOME is set:
        primary_config_dir = "$XDG_CONFIG_HOME/ccache"
    else:
        primary_config_dir = "$HOME/.config/ccache"
    primary_config_path = primary_config_dir + "/ccache.conf"

cache_dir_from_secondary_config = config.cache_dir
conf_read(conf, primary_config_path)
config.cache_dir = cache_dir_from_secondary_config  # Ignore cache_dir set in primary config

conf_update_from_environment(conf)  # Sets config.cache_dir if CCACHE_DIR is set

if config.cache_dir is not set:
    if "$HOME/.ccache" exists:
        config.cache_dir = "$HOME/.ccache"
    elif XDG_CACHE_HOME is set:
        config.cache_dir = "$XDG_CACHE_HOME/ccache"
    else:
        config.cache_dir = "$HOME/.cache/ccache"
# else: config.cache_dir was set explicitly via environment or via secondary config

# We have now determined config.cache_dir and populated the rest of config in
# prio order (1. environment, 2. primary config, 3. secondary config).
```

Comments?

**Comment by antoniojpfernandes** [comment]
> The thing is that the primary config needs to be writable since ccache -M, ccache -o, etc write to it, so it has to be associated with just one cache.

Oh, so, not having a primary config file is not an option then.

> I intended my suggested pseudo code to tie $XDG_CONFIG_HOME/ccache and $XDG_CACHE_HOME/ccache together so that if you set an explicit cache dir (either via environment or via the secondary config), $XDG_CONFIG_HOME/ccache/ccache.conf wouldn't be used at all.

Right. Thanks for clarifying the logic.

My concern was that this might be unexpected. Configurations in $XDG_CONFIG_HOME are commonly interpreted as applying whenever this specific user runs the software. But if this is made very clear in the documentation, then your solution sounds just fine to me! Thanks for giving this issue your attention.

**Comment by totph** [comment]
What about moving the default ccache directory for the upcoming v4, the storage format is incompatible anyhow. Then use os-specific locations such as `~/.cache/ccache` as mentioned here and  `~/Library/Caches` on macOS.

**Comment by jrosdahl** [comment]
> What about moving the default ccache directory for the upcoming v4, the storage format is incompatible anyhow. Then use os-specific locations such as ~/.cache/ccache as mentioned here and ~/Library/Caches on macOS.

Right, doing it in version 4 would not be bad timing, assuming you're talking about something like described in https://github.com/ccache/ccache/issues/191#issuecomment-412979701, not just changing the default. Would you be interested in working on it?

**Comment by totph** [comment]
Errr, that is quite some config logic. Here is my attempt to simplify it (somehow in even more lines!):

``` python
# config logic for ccache v4

CFG_SYSTEM = "/etc/ccache.conf" # WAS: secondary, location compiled into ccache binary
CFG_USER_HOME = default_per_user_cfg_file() # WAS: primary, respects env[XDG_CONFIG_HOME]

read_file(cfg_file) -> Config
read_env() -> Config

Config: ## [5]
    # updates own values with those present in `cfg_new`, but leaves others untouched
    update(cfg_new: Config) -> Config
    cache_dir = dir_per_user() # respects env[XDG_CACHE_HOME]
    cfg_is_final = False # only respected when set in env[CACHE_DIR]/ccache.conf
    # other cfg settings here, may be overwritten when calling `update(...)`

# init with default values from above
conf = Config()

if env[CCACHE_CONFIGPATH]:
    # Note: CCACHE_CONFIGPATH is mainly (only?) used by ccache's own test suite
    conf = conf.update(read_file("$CCACHE_CONFIGPATH"))
    # NO other config files are read, so either CCACHE_CONFIGPATH or defaults are now used
    conf = conf.update(read_env())
else:
    if env[CCACHE_DIR]: ## [1]
        dir_conf = read_file("$CCACHE_DIR/ccache.conf") ## [2]
        # ERROR if a cache_dir is set inside this config^
        dir_conf.cache_dir = "$CCACHE_DIR"
        if conf.cfg_is_final: # only these, env vars and the internal defaults are used
            conf = conf.update(dir_conf)
            conf = conf.update(read_env())
            return

    conf = conf.update(read_file(CFG_SYSTEM)) ## [4]
    conf = conf.update(read_file(CFG_USER_HOME)) ## [3]

    if dir_conf:
        conf = conf.update(dir_conf) ## [2]
    else:
        conf = conf.update(read_file(conf.cache_dir + "/ccache.conf")) ## [2]

    # conf.cache_dir now either still the default `dir_per_user()`
    # or what CCACHE_DIR, CFG_SYSTEM or CFG_USER_HOME set.

    conf = conf.update(read_env()) ## [1]

    # We have now determined config.cache_dir and populated the rest of config in
    # descending priority from:
    # 1. environment (especially CCACHE_DIR, this has an option to stop reading other configs)
    # 2. file: ccache.conf from the specified cache dir
    # 3. file: CFG_USER_HOME
    # 4. file: CFG_SYSTEM
    # 5. internal defaults
```

`CCACHE_DIR` would have an option to completely skip reading any other config file. If this is not set then the usual config order determines where the cache is placed (except for the special case of `CCACHE_CONFIGPATH`).

`~/.ccache` is completely ignored, but could be shoehorned in as step 3.5 named `CFG_USER_HOME_OLD`. But then ccache will never outgrow this location and `stat()` it with every call. _Maybe_ copy it to the new location and place a `.migrated` file there, then finally ignore it in 5 years or so.

primary and secondary was also confusing, so they have been renamed.

**Comment by jrosdahl** [comment]
> Here is my attempt to simplify it

As I understand it after a quick read-through, the diff between my and your proposal is:

1. Introduce a third configuration location (`$XDG_CONFIG_HOME/ccache`) with priority between (using the current nomenclature) the primary and the secondary configurations.
2. Store configuration inside `$XDG_CACHE_HOME/ccache` even though it is a cache directory.
3. Not tie (i.e., "store configuration for") `$XDG_CACHE_HOME/ccache` to `$XDG_CONFIG_HOME/ccache` but consider  `$XDG_CONFIG_HOME/ccache` a logically separate configuration.

Does this sound about right?

I don't immediately see why this would be simpler or better than my proposal, but that of course doesn't have to mean that it doesn't have any merits.

Questions:

1. With your proposal, what should happen when `ccache -M 10` is executed (assuming `$CCACHE_DIR` is not set)?
2. What should happen when `ccache -o umask=002` is executed?

> CCACHE_DIR would have an option to completely skip reading any other config file. 

I don't understand what you mean by this.

> But then ccache will never outgrow [~/.ccache] and stat() it with every call.

I don't see why that would be a problem.

>  _Maybe_ copy it to the new location and place a .migrated file there, then finally ignore it in 5 years or so.

"Copy" as in "copy the whole cache directory"? That doesn't sound feasible.

> primary and secondary was also confusing, so they have been renamed.

Confusing after introducing a third config place or confusing even before? If the latter, why?

**Comment by totph** [comment]
> 2. Store configuration inside `$XDG_CACHE_HOME/ccache` even though it is a cache directory.

Right, I overlooked that. Removing `~/.cache` should not remove configuration data. `$XDG_CONFIG_HOME/ccache/..` (or just `ccache.conf`?) should indeed be tied to the above cache location (as mentioned in 3.).

>  1. With your proposal, what should happen when `ccache -M 10` is executed (assuming `$CCACHE_DIR` is not set)?

It gets written to `$XDG_CONFIG_HOME/ccache`.

 >  2. What should happen when `ccache -o umask=002` is executed?

The same as before.

>>  CCACHE_DIR would have an option to completely skip reading any other config file.
>
> I don't understand what you mean by this.

At the moment if that env var is set `$CCACHE_DIR/ccache.conf` is read, then the primary and secondary configs. These might set undesired extra values. This could be prevented by an option which sets the above config as the exclusive and only one. This way very specific settings for just one project can be ensured, e.g. in a build script which sets `CCACHE_DIR` to `./build-cache` and does not want user or system config settings having any effect.

> > _Maybe_ copy it to the new location and place a .migrated file there, then finally ignore it in 5 years or so.
>
> "Copy" as in "copy the whole cache directory"? That doesn't sound feasible.

If `~/.ccache/is_migrated` exists, ignore it. If not and `~/.cache/ccache.conf` exists but `~/.config/ccache'` not yet, then copy just the config over and start storing new data in `~/.cache/ccache`.

But now that I think about it, just checking for `~/.cache` and continue using that for data and config, and only if it is missing switch to `~/.cache/ccache/..` / `~/.config` is just as good. Maybe even better for people who already have specific cleanup scripts looking at `~/.ccache`.

> > primary and secondary was also confusing, so they have been renamed.
> 
> Confusing after introducing a third config place or confusing even before? If the latter, why?

The latter. Naming them according to the location where they are placed seems easier to understand. E.g. my intuition for `primary` was the system location under `/etc/`. Maybe go with `primary_user_cfg`, `secondary_system_wide_cfg`?

**Comment by xerus2000** [comment]
Please reconsider this, I want my HOME clean :)

**Comment by jrosdahl** [comment]
@xerus2000 wrote:
> Please reconsider this, I want my HOME clean :)

Hm... Either you think that using XDG_CONFIG_HOME makes your HOME less clean (if so, why?) or you think that we have decided to not use XDG_CONFIG_HOME (if so, why?). :slightly_smiling_face:

**Comment by antoniojpfernandes** [comment]
Great! Thanks for your work. It's appreciated. As a fellow FOSS maintainer I know 3 years has been quite fast for this kind of issue.

**Comment by xerus2000** [comment]
@jrosdahl oh, I am on 3.4.1, that's why I thought this was still not solved. And from skimming through the issue I didn't pick up the progress. Thank you very much :)

**Comment by xerus2000** [comment]
Am I correctly seeing that it will only be available from 4.0, which isn't yet released?

**Comment by jrosdahl** [comment]
> Am I correctly seeing that it will only be available from 4.0, which isn't yet released?

Yes. Issues marked with milestone 4.0 will be included in version 4.0.
